vault/diagnose/tls_verification.go

```diff
@@ -0,0 +1,138 @@
+package diagnose
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/hashicorp/vault/internalshared/listenerutil"
+	"github.com/hashicorp/vault/sdk/helper/tlsutil"
+	"github.com/hashicorp/vault/vault"
+)
+
+const minVersionError = "'tls_min_version' value %q not supported, please specify one of [tls10,tls11,tls12,tls13]"
+const maxVersionError = "'tls_max_version' value %q not supported, please specify one of [tls10,tls11,tls12,tls13]"
+
+func ListenerChecks(listeners []listenerutil.Listener) error {
+	for _, listener := range listeners {
+		l := listener.Config
+
+		// Perform the TLS version check for listeners.
+		if l.TLSMinVersion == "" {
+			l.TLSMinVersion = "tls12"
+		}
+		if l.TLSMaxVersion == "" {
+			l.TLSMaxVersion = "tls13"
+		}
+		_, ok := tlsutil.TLSLookup[l.TLSMinVersion]
+		if !ok {
+			return fmt.Errorf(minVersionError, l.TLSMinVersion)
+		}
+		_, ok = tlsutil.TLSLookup[l.TLSMaxVersion]
+		if !ok {
+			return fmt.Errorf(maxVersionError, l.TLSMaxVersion)
+		}
+
+		var err error
+		// Perform checks on the TLS Cryptographic Information.
+		if err = TLSFileChecks(l.TLSCertFile, l.TLSKeyFile); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// TLSFileChecks contains manual error checks against the TLS configuration
+func TLSFileChecks(certFilePath, keyFilePath string) error {
```

Comment:
Modularize this to load the cert from a file path separately, and have the checks in a separate function so that we can load from urls as well.